packages/lit-element/src/lib/platform-support.ts

```diff
@@ -0,0 +1,388 @@
+
+  const scopedTemplateCache = new Map<
+    string,
+    Map<TemplateStringsArray, TemplateStandIn>
+  >();
+
+  // Note, it's ok to subclass Template since it's only used via NodePart.
+  let currentResult: ShadyTemplateResult;
+  class ShadyTemplate extends Template {
+    constructor(result: ShadyTemplateResult) {
+      // TODO(sorvell): Decide if Template should store the result to avoid this.
+      // We need the `result` in `_createElement` which is called directly
+      // from the constructor. Since we can't access this before super, uce
+      // a side channel to store the data.
+      currentResult = result;
+      super(result);
+    }
+    /**
+     * Override to extract style elements from the template
+     * and store all style.textContent in the shady scope data.
+     */
+    _createElement(html: string) {
+      const template = super._createElement(html);
+      if (!window.ShadyCSS!.nativeShadow) {
+        window.ShadyCSS!.prepareTemplateDom(template, currentScope);
+      }
+      const scopeCss = cssForScope(currentScope);
+      if (scopeCss !== undefined) {
+        // Remove styles and store textContent.
+        const styles = template.content.querySelectorAll('style') as NodeListOf<
+          HTMLStyleElement
+        >;
+        scopeCss.push(
+          ...Array.from(styles).map((style) => {
+            style.parentNode?.removeChild(style);
+            return style.textContent!;
+          })
+        );
+        // Apply styles if this is the result rendering into shadowRoot.
+        if (currentResult.__renderIntoShadowRoot) {
+          if (scopeCss.length) {
+            const style = document.createElement('style');
+            style.textContent = scopeCss.join('\n');
+            // Place aggregate style into template. Note, ShadyCSS will remove
+            // it if polyfilled ShadowDOM is used.
+            template.content.insertBefore(style, template.content.firstChild);
+          }
+          ensureStyles(currentScope, template);
+        }
+      }
+      return template;
+    }
+  }
+
+  Object.assign(NodePart.prototype, {
+    /**
+     * Patch to apply gathered css via ShadyCSS. This is done only once per scope.
+     */
+    __baseSetValue: NodePart.prototype._setValue,
+    _setValue(this: NodePartStandIn, value: unknown) {
+      const container = this._startNode.parentNode!;
+      const renderingIntoShadowRoot = container instanceof ShadowRoot;
+      if (!renderingIntoShadowRoot) {
+        this.__baseSetValue(value);
+      } else {
+        const previousScope = currentScope;
+        currentScope = this.options.scope ?? '';
+
+        const renderingTemplateResult = !!(value as ShadyTemplateResult)
+          ?._$litType$;
+        // If result is a TemplateResult decorate it so that when the template
+        // is created, styling can be shimmed.
+        if (renderingTemplateResult) {
+          (value as ShadyTemplateResult).__renderIntoShadowRoot = renderingIntoShadowRoot;
```

Comment:
If we're going to set this (always true), can be the scope instead and avoid one of the side-channel variables (`currentScope`)?

Reply:
I misunderstood, this is marking the _top_ template, but nested templates still need the `currentScope`.